plugins/Phaser/PhaserControls.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * PhaserControls.cpp
+ *
+ * Copyright (c) 2019 Lost Robot <r94231@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+
+#include <QDomElement>
+
+#include "PhaserControls.h"
+#include "PhaserEffect.h"
+#include "Engine.h"
+#include "Song.h"
+#include "GuiApplication.h"
+#include "MainWindow.h"
+
+
+const int PHA_DOT_SLIDER_LENGTH = 338;
+const float PHA_MIN_FREQ = 20;
+const float PHA_MAX_FREQ = 20000;
+
+
+PhaserControls::PhaserControls( PhaserEffect* effect ) :
+	EffectControls( effect ),
+	m_effect( effect ),
+	m_cutoffModel( 640.0f, 20.0f, 20000.0f, 0.01f, this, tr( "Cutoff" ) ),
+	m_resonanceModel( 0.05f, 0.05f, 2.0f, 0.001f, this, tr( "Resonance" ) ),
+	m_feedbackModel( 0.0f, -100.0f, 100.0f, 0.01f, this, tr( "Feedback" ) ),
+	m_orderModel(8, 1, 32, this, tr("Order")),
+	m_delayModel( 1.0f, 1.0f, 50.0f, 1.0f, this, tr( "Delay" ) ),
+	m_rateModel( 10.0f, 0.01f, 60.0f, 0.001f, 60000.0, this, tr( "Rate" ) ),
+	m_amountModel( 5.0f, 0.f, 5.0f, 0.01f, this, tr( "Amount" ) ),
+	m_enableLFOModel( true , this, tr( "Enable LFO" ) ),
+	m_phaseModel( 180.0f, 0.f, 360.0f, 0.1f, this, tr( "Phase" ) ),
+	m_wetDryModel( 0.5f, 0.f, 1.0f, 0.01f, this, tr( "Wet/Dry" ) ),
+	m_inFollowModel( 0.f, -15.f, 15.0f, 0.01f, this, tr( "Input Follow" ) ),
+	m_attackModel( 500.0f, 0.0f, 2000.0f, 1.0f, this, tr( "Attack" ) ),
+	m_releaseModel( 500.0f, 0.0f, 2000.0f, 1.0f, this, tr( "Release" ) ),
+	m_outGainModel( 0.0, -60.0, 20.0, 0.01, this, tr( "Output gain" ) ),
+	m_inGainModel( 0.0, -60.0, 20.0, 0.01, this, tr( "Input gain" ) )
+{
+	m_cutoffModel.setScaleLogarithmic(true);
+	m_resonanceModel.setScaleLogarithmic(true);
+	m_delayModel.setScaleLogarithmic(true);
+
+	connect( Engine::mixer(), SIGNAL( sampleRateChanged() ), this, SLOT( changedSampleRate() ) );
+	connect( Engine::getSong(), SIGNAL( playbackStateChanged() ), this, SLOT( changedPlaybackState() ) );
+
+	connect( &m_phaseModel, SIGNAL( dataChanged() ), this, SLOT( updatePhase() ) );
+
+	m_outPeakL = 0.0;
+	m_outPeakR = 0.0;
+}
+
+
+PhaserControls::~PhaserControls()
+{
+	disconnect( gui->mainWindow(), SIGNAL( periodicUpdate() ), 0, 0 );
```

Comment:
Not needed anymore.

Reply:
Whoa!  How did that sneak its way into there?  I'm sure I deleted that.  Weird.